misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -0,0 +1,117 @@
+  private val config: AwsSqsJobQueueConfig,
+  private val sqs: AmazonSQS,
+  private val queueUrls: QueueUrlMapping,
+  @ForSqsConsumer private val dispatchThreadPool: ExecutorService,
+  private val tracer: Tracer,
+  private val metrics: SqsMetrics
+) : AbstractIdleService(), JobConsumer {
+  private val subscriptions = ConcurrentHashMap<QueueName, JobConsumer.Subscription>()
+
+  override fun startUp() {}
+
+  override fun shutDown() {
+    subscriptions.values.forEach { it.close() }
+  }
+
+  override fun subscribe(queueName: QueueName, handler: JobHandler): JobConsumer.Subscription {
+    val receiver = QueueReceiver(queueName, handler)
+    if (subscriptions.putIfAbsent(queueName, receiver) != null) {
+      throw IllegalStateException("already subscribed to queue ${queueName.value}")
+    }
+
+    log.info {
+      "subscribing to queue ${queueName.value} with ${config.concurrent_receivers_per_queue} receivers"
+    }
+
+    for (i in (0 until config.concurrent_receivers_per_queue)) {
+      thread(name = "sqs-receiver-${queueName.value}-$i", start = true) {
+        log.info { "launching receiver $i for ${queueName.value}" }
+        receiver.run()
+      }
+    }
+
+    return receiver
+  }
+
+  private inner class QueueReceiver(
+    private val queue: QueueName,
+    private val handler: JobHandler
+  ) : Runnable, JobConsumer.Subscription {
+
+    private val queueUrl = queueUrls[queue]
+    private val running = AtomicBoolean(true)
+
+    override fun run() {
+      while (running.get()) {
+        val messages = sqs.receiveMessage(ReceiveMessageRequest()
+            .withAttributeNames("All")
+            .withMessageAttributeNames("All")
+            .withQueueUrl(queueUrl)
+            .withMaxNumberOfMessages(10))
+            .messages
+
+        messages.map { SqsJob(queue, queueUrls, sqs, metrics, it) }.forEach { message ->
+          dispatchThreadPool.submit {
+            metrics.jobsReceived.labels(queue.value).inc()
+
+            tracer.traceWithSpan("handle-job-${queue.value}") { span ->
+              // If the incoming job has an original trace id, set that as a tag on the new span.
+              // We don't turn that into the parent of the current span because that would
+              // incorrectly include the execution time of the job in the execution time of the
+              // action that triggered the job
+              message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
+                ORIGINAL_TRACE_ID_TAG.set(span, it)
+              }
+
+              // Run the handler and record timing
+              try {
+                val (duration, _) = timed { handler.handleJob(message) }
+                metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queue.value)
+              } catch (th: Throwable) {
+                log.error(th) { "error handling job from ${queue.value}" }
+                metrics.handlerFailures.labels(queue.value).inc()
```

Comment:
we probably want to include the shard number as a label, so we can identify slow/stuck receivers

Reply:
Not sure what this means. There's no real "shard" per se - just a pool of receivers. Which messages the receivers receive have no sharding implications, and the receiver doesn't wait for the current batch to finish dispatch before asking for more (unless the dispatch thread pool is full and applying backpressure)

Reply:
k. yea per-thread metrics is probably not interesting